File: core/src/test/kotlin/dev/shelly/hermes/core/AgentCoreTest.kt
package dev.shelly.hermes.core

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertIs
import kotlin.test.assertTrue

class AgentCoreTest {
    @Test
    fun `executes approved tool then completes`() = kotlinx.coroutines.test.runTest {
        var calls = 0
        val saved = mutableListOf<AgentCheckpoint>()
        val core = AgentCore(
            model = ModelGateway {
                if (calls++ == 0) ModelReply(toolCalls = listOf(ToolCall("1", "read_file", "{}")))
                else ModelReply(content = "完成")
            },
            tools = ToolExecutor { "content" },
            approvals = ApprovalGateway { ApprovalDecision.APPROVE },
            checkpoints = CheckpointStore { saved += it }
        )
        val result = core.run(listOf(AgentMessage(MessageRole.USER, "检查项目")), object : CancellationSignal {
            override val isCancelled = false
        })
        assertIs<AgentResult.Completed>(result)
        assertEquals("完成", result.message)
        assertEquals(1, result.checkpoint.toolCalls)
    }

    @Test
    fun `stops before model call when cancelled`() = kotlinx.coroutines.test.runTest {
        val core = AgentCore(
            model = ModelGateway { error("must not run") },
            tools = ToolExecutor { "" },
            approvals = ApprovalGateway { ApprovalDecision.APPROVE },
            checkpoints = CheckpointStore { }
        )
        val result = core.run(emptyList(), object : CancellationSignal {
            override val isCancelled = true
        })
        assertEquals("cancelled", assertIs<AgentResult.Stopped>(result).reason)
    }

    @Test
    fun `default policy still requests approval for read only tools`() = kotlinx.coroutines.test.runTest {
        var modelCalls = 0
        var approvalCalls = 0
        val core = AgentCore(
            model = ModelGateway {
                if (modelCalls++ == 0) ModelReply(toolCalls = listOf(ToolCall("read-1", "read_file", "{}")))
                else ModelReply(content = "done")
            },
            tools = ToolExecutor { "content" },
            approvals = ApprovalGateway {
                approvalCalls += 1
                ApprovalDecision.APPROVE
            },
            checkpoints = CheckpointStore { }
        )

        assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled()))
        assertEquals(1, approvalCalls)
    }

    @Test
    fun `read only policy skips approval round trip`() = kotlinx.coroutines.test.runTest {
        var modelCalls = 0
        val executed = mutableListOf<String>()
        val core = AgentCore(
            model = ModelGateway {
                if (modelCalls++ == 0) {
                    ModelReply(toolCalls = listOf(
                        ToolCall("read-1", "read_file", "{}"),
                        ToolCall("exists-1", "exists", "{}")
                    ))
                } else ModelReply(content = "done")
            },
            tools = ToolExecutor { call -> executed += call.name; "ok" },
            approvals = ApprovalGateway { error("read-only tools must not wait for approval") },
            checkpoints = CheckpointStore { },
            approvalPolicy = ToolApprovalPolicy.autoApproveReadOnly()
        )

        assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled()))
        assertEquals(listOf("read_file", "exists"), executed)
    }

    @Test
    fun `read only policy continues to require approval for mutations`() = kotlinx.coroutines.test.runTest {
        var modelCalls = 0
        var approvalCalls = 0
        val core = AgentCore(
            model = ModelGateway {
                if (modelCalls++ == 0) ModelReply(toolCalls = listOf(ToolCall("write-1", "overwrite_file", "{}")))
                else ModelReply(content = "done")
            },
            tools = ToolExecutor { "written" },
            approvals = ApprovalGateway {
                approvalCalls += 1
                ApprovalDecision.APPROVE
            },
            checkpoints = CheckpointStore { },
            approvalPolicy = ToolApprovalPolicy.autoApproveReadOnly()
        )

        assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled()))
        assertEquals(1, approvalCalls)
    }

    @Test
    fun `broken approval policy fails closed`() = kotlinx.coroutines.test.runTest {
        var modelCalls = 0
        var approvalCalls = 0
        val core = AgentCore(
            model = ModelGateway {
                if (modelCalls++ == 0) ModelReply(toolCalls = listOf(ToolCall("read-1", "read_file", "{}")))
                else ModelReply(content = "done")
            },
            tools = ToolExecutor { "content" },
            approvals = ApprovalGateway {
                approvalCalls += 1
                ApprovalDecision.APPROVE
            },
            checkpoints = CheckpointStore { },
            approvalPolicy = ToolApprovalPolicy { error("invalid policy state") }
        )

        assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled()))
        assertEquals(1, approvalCalls)
    }

    @Test
    fun `observer reports model approval and tool lifecycle with durations`() = kotlinx.coroutines.test.runTest {
        var modelCalls = 0
        var clock = 0L
        val events = mutableListOf<AgentEvent>()
        val core = AgentCore(
            model = ModelGateway {
                if (modelCalls++ == 0) ModelReply(toolCalls = listOf(ToolCall("write-1", "overwrite_file", "{}")))
                else ModelReply(content = "done")
            },
            tools = ToolExecutor { "written" },
            approvals = ApprovalGateway { ApprovalDecision.APPROVE },
            checkpoints = CheckpointStore { },
            observer = AgentObserver { events += it },
            nanoTime = { clock.also { clock += 2_000_000L } }
        )

        assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled()))

        assertEquals(AgentEvent.ModelStarted(1), events[0])
        assertEquals(2L, assertIs<AgentEvent.ModelFinished>(events[1]).durationMillis)
        assertIs<AgentEvent.ApprovalWaiting>(events[2])
        assertEquals(2L, assertIs<AgentEvent.ApprovalFinished>(events[3]).durationMillis)
        val toolStarted = assertIs<AgentEvent.ToolStarted>(events[4])
        assertEquals("overwrite_file", toolStarted.toolName)
        val toolFinished = assertIs<AgentEvent.ToolFinished>(events[5])
        assertEquals("overwrite_file", toolFinished.toolName)
        assertEquals(2L, toolFinished.durationMillis)
        assertTrue(toolFinished.succeeded)
        assertIs<AgentEvent.ModelStarted>(events[6])
        assertIs<AgentEvent.ModelFinished>(events[7])
    }

    @Test
    fun `streaming gateway emits model deltas before final reply`() = kotlinx.coroutines.test.runTest {
        val events = mutableListOf<AgentEvent>()
        val core = AgentCore(
            model = StreamingGateway(listOf("Hello ", "world")),
            tools = ToolExecutor { "unused" },
            approvals = ApprovalGateway { error("no tools requested") },
            checkpoints = CheckpointStore { },
            observer = AgentObserver { events += it },
        )

        val result = assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled()))
        assertEquals("Hello world", result.message)
        val deltaTexts = events.filterIsInstance<AgentEvent.ModelDelta>().map(AgentEvent.ModelDelta::text)
        assertEquals(listOf("Hello ", "world"), deltaTexts)
    }

    @Test
    fun `failed tool still emits completion event`() = kotlinx.coroutines.test.runTest {
        val events = mutableListOf<AgentEvent>()
        val core = AgentCore(
            model = ModelGateway { ModelReply(toolCalls = listOf(ToolCall("bad-1", "overwrite_file", "{}"))) },
            tools = ToolExecutor { error("disk failed") },
            approvals = ApprovalGateway { ApprovalDecision.APPROVE },
            checkpoints = CheckpointStore { },
            observer = AgentObserver { events += it }
        )

        val failure = runCatching { core.run(emptyList(), neverCancelled()) }
        assertTrue(failure.isFailure)
        val finished = assertIs<AgentEvent.ToolFinished>(events.last())
        assertFalse(finished.succeeded)
        assertEquals("overwrite_file", finished.toolName)
    }

    @Test
    fun `observer failure does not interrupt execution`() = kotlinx.coroutines.test.runTest {
        val core = AgentCore(
            model = ModelGateway { ModelReply(content = "done") },
            tools = ToolExecutor { "unused" },
            approvals = ApprovalGateway { ApprovalDecision.APPROVE },
            checkpoints = CheckpointStore { },
            observer = AgentObserver { error("telemetry failure") }
        )

        val result = assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled()))
        assertEquals("done", result.message)
    }

    @Test
    fun `resume approved but unstarted tool call does not prompt again`() = kotlinx.coroutines.test.runTest {
        var approvalCalls = 0
        val executed = mutableListOf<String>()
        val core = AgentCore(
            model = ModelGateway { ModelReply(content = "done") },
            tools = ToolExecutor { call -> executed += call.name; "written" },
            approvals = ApprovalGateway { approvalCalls += 1; ApprovalDecision.APPROVE },
            checkpoints = CheckpointStore { },
        )
        val checkpoint = AgentCheckpoint(
            messages = listOf(AgentMessage(MessageRole.ASSISTANT, "", toolCalls = listOf(ToolCall("write-1", "overwrite_file", "{}")))),
            round = 1,
            consumedTokens = 0,
            toolCalls = 1,
            pendingToolCalls = listOf(PendingToolCall(ToolCall("write-1", "overwrite_file", "{}"), ToolExecutionStage.AWAITING_EXECUTION)),
        )

        assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled(), checkpoint))
        assertEquals(0, approvalCalls)
        assertEquals(listOf("overwrite_file"), executed)
    }

    @Test
    fun `resume interrupted running mutation asks user before retry`() = kotlinx.coroutines.test.runTest {
        var approvalCalls = 0
        val executed = mutableListOf<String>()
        val core = AgentCore(
            model = ModelGateway { ModelReply(content = "done") },
            tools = ToolExecutor { call -> executed += call.name; "written again" },
            approvals = ApprovalGateway {
                approvalCalls += 1
                ApprovalDecision.APPROVE
            },
            checkpoints = CheckpointStore { },
        )
        val checkpoint = AgentCheckpoint(
            messages = emptyList(),
            round = 1,
            consumedTokens = 0,
            toolCalls = 1,
            pendingToolCalls = listOf(PendingToolCall(ToolCall("write-1", "overwrite_file", "{}"), ToolExecutionStage.RUNNING)),
        )

        assertIs<AgentResult.Completed>(core.run(emptyList(), neverCancelled(), checkpoint))
        assertEquals(1, approvalCalls)
        assertEquals(listOf("overwrite_file"), executed)
    }

    private fun neverCancelled() = object : CancellationSignal {
        override val isCancelled = false
    }

    private class StreamingGateway(
        private val chunks: List<String>,
    ) : ModelGateway, StreamingModelGateway {
        override suspend fun complete(messages: List<AgentMessage>): ModelReply =
            ModelReply(content = chunks.joinToString(""))

        override suspend fun completeStreaming(
            messages: List<AgentMessage>,
            onDelta: (String) -> Unit,
        ): ModelReply {
            chunks.forEach(onDelta)
            return ModelReply(content = chunks.joinToString(""))
        }
    }
}

File: flutter/test/core/agent_core_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:shelly_hermes/core/agent_core.dart';
import 'package:shelly_hermes/core/approval_broker.dart';
import 'package:shelly_hermes/core/models.dart';

/// Scripted model: replays queued replies, then keeps returning the last one.
class FakeModel implements ModelGateway {
  final List<ModelReply> script;
  final List<List<AgentMessage>> calls = [];

  FakeModel(this.script);

  @override
  Future<ModelReply> complete(List<AgentMessage> messages) async {
    calls.add(List.of(messages));
    return script.length <= 1 ? script.first : script.removeAt(0);
  }
}

class FakeTools implements ToolExecutor {
  final Map<String, String> outcomes = {};
  final List<ToolCall> executed = [];
  Object? throwOn;

  @override
  Future<String> execute(ToolCall call) async {
    if (throwOn != null) throw throwOn!;
    executed.add(call);
    return outcomes[call.name] ?? 'ok:${call.name}';
  }
}

/// Approval gateway resolving from a queue; empty queue means approve all.
class QueueApprovals implements ApprovalGateway {
  final List<ApprovalDecision> decisions;
  final List<ToolCall> requests = [];

  QueueApprovals([this.decisions = const []]);

  @override
  Future<ApprovalDecision> request(ToolCall call) async {
    requests.add(call);
    return decisions.isEmpty ? ApprovalDecision.approve : decisions.removeAt(0);
  }
}

class MemoryCheckpoints implements CheckpointStore {
  final List<AgentCheckpoint> saved = [];

  @override
  Future<void> save(AgentCheckpoint checkpoint) async => saved.add(checkpoint);
}

AgentMessage userMsg(String text) => AgentMessage(role: MessageRole.user, content: text);

void main() {
  test('completes when the model returns plain content', () async {
    final model = FakeModel([const ModelReply(content: '你好,世界')]);
    final checkpoints = MemoryCheckpoints();
    final core = AgentCore(
      model: model,
      tools: FakeTools(),
      approvals: QueueApprovals(),
      checkpoints: checkpoints,
    );

    final result = await core.run([userMsg('hi')], CancelFlag());

    expect(result, isA<AgentCompleted>());
    expect((result as AgentCompleted).message, '你好,世界');
    expect(result.checkpoint.messages.last.role, MessageRole.assistant);
    expect(checkpoints.saved, isNotEmpty);
  });

  test('executes an approved tool call and feeds the result back', () async {
    final model = FakeModel([
      const ModelReply(content: '', toolCalls: [
        ToolCall(id: 't1', name: 'list_files', argumentsJson: '{}'),
      ]),
      const ModelReply(content: 'done'),
    ]);
    final tools = FakeTools()..outcomes['list_files'] = 'a.txt\nb.txt';
    final approvals = QueueApprovals();
    final core = AgentCore(
      model: model,
      tools: tools,
      approvals: approvals,
      checkpoints: MemoryCheckpoints(),
    );

    final result = await core.run([userMsg('列出文件')], CancelFlag());

    expect((result as AgentCompleted).message, 'done');
    expect(tools.executed.single.name, 'list_files');
    expect(approvals.requests.single.id, 't1');
    final toolMessage = model.calls[1].last;
    expect(toolMessage.role, MessageRole.tool);
    expect(toolMessage.content, 'a.txt\nb.txt');
    expect(toolMessage.toolCallId, 't1');
  });

  test('rejected tool call records a rejection tool message without executing',
      () async {
    final model = FakeModel([
      const ModelReply(content: '', toolCalls: [
        ToolCall(id: 't1', name: 'write_file', argumentsJson: '{}'),
      ]),
      const ModelReply(content: 'understood'),
    ]);
    final tools = FakeTools();
    final core = AgentCore(
      model: model,
      tools: tools,
      approvals: QueueApprovals([ApprovalDecision.reject]),
      checkpoints: MemoryCheckpoints(),
    );

    await core.run([userMsg('hi')], CancelFlag());

    expect(tools.executed, isEmpty);
    final toolMessage = model.calls[1].last;
    expect(toolMessage.role, MessageRole.tool);
    expect(toolMessage.content, 'Tool call rejected by user');
  });

  test('apply_patch approval expands into per-hunk calls and collapses the '
      'approved subset', () async {
    final model = FakeModel([
      const ModelReply(content: '', toolCalls: [
        ToolCall(
          id: 'p1',
          name: 'apply_patch',
          argumentsJson: '{"path":"lib/a.dart","patch":"@@ -1,2 +1,2 @@\\n-a\\n+b\\n@@ -3,2 +3,2 @@\\n-c\\n+d"}',
        ),
      ]),
      const ModelReply(content: 'patched'),
    ]);
    final tools = FakeTools();
    final approvals = QueueApprovals([
      ApprovalDecision.approve,
      ApprovalDecision.reject,
    ]);
    final core = AgentCore(
      model: model,
      tools: tools,
      approvals: approvals,
      checkpoints: MemoryCheckpoints(),
    );

    await core.run([userMsg('apply')], CancelFlag());

    // Two hunks requested individually; user approved only the first.
    expect(approvals.requests.map((r) => r.id), ['p1:hunk-1', 'p1:hunk-2']);
    expect(tools.executed.single.name, 'apply_patch');
    expect(tools.executed.single.id, 'p1');
    // Only the approved hunk survives in the collapsed patch.
    expect(tools.executed.single.argumentsJson, isNot(contains('-c')));
  });

  test('read-only policy bypasses approval for trusted tools', () async {
    final model = FakeModel([
      const ModelReply(content: '', toolCalls: [
        ToolCall(id: 't1', name: 'read_file', argumentsJson: '{}'),
      ]),
      const ModelReply(content: 'done'),
    ]);
    final approvals = QueueApprovals();
    final core = AgentCore(
      model: model,
      tools: FakeTools(),
      approvals: approvals,
      checkpoints: MemoryCheckpoints(),
      approvalPolicy: AutoApproveReadOnlyPolicy(),
    );

    await core.run([userMsg('hi')], CancelFlag());

    expect(approvals.requests, isEmpty);
  });

  test('broken approval policy fails closed to requiring approval', () async {
    final model = FakeModel([
      const ModelReply(content: '', toolCalls: [
        ToolCall(id: 't1', name: 'read_file', argumentsJson: '{}'),
      ]),
      const ModelReply(content: 'done'),
    ]);
    final approvals = QueueApprovals();
    final core = AgentCore(
      model: model,
      tools: FakeTools(),
      approvals: approvals,
      checkpoints: MemoryCheckpoints(),
      approvalPolicy: _ThrowingPolicy(),
    );

    await core.run([userMsg('hi')], CancelFlag());

    // Fail closed: the request was still routed to the human gateway.
    expect(approvals.requests, isNotEmpty);
  });

  test('cancellation before a round returns stopped with checkpoint',
      () async {
    final core = AgentCore(
      model: FakeModel([const ModelReply(content: 'x')]),
      tools: FakeTools(),
      approvals: QueueApprovals(),
      checkpoints: MemoryCheckpoints(),
    );
    final flag = CancelFlag()..cancel();

    final result = await core.run([userMsg('hi')], flag);

    expect(result, isA<AgentStopped>());
    expect((result as AgentStopped).reason, 'cancelled');
  });

  test('round limit returns stopped', () async {
    final model = FakeModel([
      const ModelReply(content: '', toolCalls: [
        ToolCall(id: 't1', name: 'noop', argumentsJson: '{}'),
      ]),
    ]);
    final core = AgentCore(
      model: model,
      tools: FakeTools(),
      approvals: QueueApprovals(),
      checkpoints: MemoryCheckpoints(),
      limits: const AgentLimits(maxRounds: 2, maxToolCalls: 100),
    );

    final result = await core.run([userMsg('hi')], CancelFlag());

    expect((result as AgentStopped).reason, 'round_limit_exceeded');
  });

  test('token budget exceeded returns stopped', () async {
    final core = AgentCore(
      model: FakeModel([const ModelReply(content: 'x', inputTokens: 70000)]),
      tools: FakeTools(),
      approvals: QueueApprovals(),
      checkpoints: MemoryCheckpoints(),
    );

    final result = await core.run([userMsg('hi')], CancelFlag());

    expect((result as AgentStopped).reason, 'token_budget_exceeded');
  });

  test('resume from awaitingExecution executes without re-asking approval',
      () async {
    const pendingCall = ToolCall(id: 't9', name: 'write_file', argumentsJson: '{}');
    final resume = AgentCheckpoint(
      messages: [userMsg('hi'), const AgentMessage(role: MessageRole.assistant, content: '', toolCalls: [pendingCall])],
      round: 1,
      consumedTokens: 10,
      toolCalls: 1,
      pendingToolCalls: const [
        PendingToolCall(call: pendingCall, stage: ToolExecutionStage.awaitingExecution),
      ],
    );
    final model = FakeModel([const ModelReply(content: 'resumed-done')]);
    final approvals = QueueApprovals();
    final core = AgentCore(
      model: model,
      tools: FakeTools()..outcomes['write_file'] = 'written',
      approvals: approvals,
      checkpoints: MemoryCheckpoints(),
    );

    final result = await core.run([userMsg('hi')], CancelFlag(), resumeFrom: resume);

    expect(approvals.requests, isEmpty);
    expect((result as AgentCompleted).message, 'resumed-done');
    expect(model.calls.single.last.content, 'written');
  });

  test('resume with RUNNING stage demotes to awaitingApproval and asks again',
      () async {
    const pendingCall = ToolCall(id: 't9', name: 'write_file', argumentsJson: '{}');
    final resume = AgentCheckpoint(
      messages: const [],
      round: 1,
      consumedTokens: 10,
      toolCalls: 1,
      pendingToolCalls: [
        PendingToolCall(call: pendingCall, stage: ToolExecutionStage.running),
      ],
    );
    final model = FakeModel([const ModelReply(content: 'done')]);
    final approvals = QueueApprovals([ApprovalDecision.approve]);
    final core = AgentCore(
      model: model,
      tools: FakeTools(),
      approvals: approvals,
      checkpoints: MemoryCheckpoints(),
    );

    await core.run(const [], CancelFlag(), resumeFrom: resume);

    // The human must confirm explicitly; execution is not silently rerun.
    expect(approvals.requests.single.id, 't9');
  });

  test('observer receives lifecycle events and cannot break the loop',
      () async {
    final events = <AgentEvent>[];
    final core = AgentCore(
      model: FakeModel([const ModelReply(content: 'hi there')]),
      tools: FakeTools(),
      approvals: QueueApprovals(),
      checkpoints: MemoryCheckpoints(),
      observer: _ThrowingObserver(events),
    );

    await core.run([userMsg('hi')], CancelFlag());

    expect(events.whereType<ModelStarted>(), isNotEmpty);
    expect(events.whereType<ModelFinished>(), isNotEmpty);
  });

  test('streaming gateway deltas are surfaced as ModelDelta events', () async {
    final events = <AgentEvent>[];
    final core = AgentCore(
      model: _StreamingFakeModel(),
      tools: FakeTools(),
      approvals: QueueApprovals(),
      checkpoints: MemoryCheckpoints(),
      observer: _CollectingObserver(events),
    );

    final result = await core.run([userMsg('hi')], CancelFlag());

    expect((result as AgentCompleted).message, 'streamed');
    expect(
      events.whereType<ModelDelta>().map((e) => e.text).join(),
      'stred',
    );
  });
}

class _ThrowingPolicy implements ToolApprovalPolicy {
  @override
  bool requiresApproval(ToolCall call) => throw StateError('broken policy');
}

class _ThrowingObserver implements AgentObserver {
  _ThrowingObserver(this.events);

  final List<AgentEvent> events;

  @override
  void onEvent(AgentEvent event) {
    events.add(event);
    throw StateError('observer exploded');
  }
}

class _CollectingObserver implements AgentObserver {
  _CollectingObserver(this.events);

  final List<AgentEvent> events;

  @override
  void onEvent(AgentEvent event) => events.add(event);
}

class _StreamingFakeModel implements StreamingModelGateway {
  @override
  Future<ModelReply> complete(List<AgentMessage> messages) async =>
      const ModelReply(content: 'streamed');

  @override
  Future<ModelReply> completeStreaming(
    List<AgentMessage> messages,
    void Function(String) onDelta,
  ) async {
    onDelta('str');
    onDelta('e');
    onDelta('d');
    return const ModelReply(content: 'streamed');
  }
}
